packages/web/integrations/axelar/hooks/use-deposit-address.ts

```diff
@@ -15,96 +28,66 @@ export function useDepositAddress(
+) {
+  const [axelarAssetTransfer] = useState(
+    () => new AxelarAssetTransfer({ environment })
+  );
 
-  /** Key: sourceChain/destChain/address/coinMinimalDenom/autoUnwrap */
-  const depositAddressCache = useRef(new Map<string, string>());
-  /** Remembers most recent generating address. */
-  const latestGenCacheKey = useRef("");
+  const {
+    get: cacheGet,
+    set: cacheSet,
+    has: cacheHas,
+  } = useCache<IPromiseBasedObservable<string>>();
 
-  const generateAddress = useCallback(async () => {
+  const observable = useMemo(() => {
     const cacheKey = `${sourceChain}/${destChain}/${destinationAddress}/${coinMinimalDenom}/${Boolean(
-      autoUnwrapIntoNative
+      shouldUnwrapIntoNative
     )}`;
-    const cachedDepositAddress = depositAddressCache.current.get(cacheKey);
-    if (cachedDepositAddress) {
-      setDepositAddress(cachedDepositAddress);
-    } else if (destinationAddress) {
-      setIsLoading(true);
-      latestGenCacheKey.current = cacheKey;
-      new AxelarAssetTransfer({ environment })
-        .getDepositAddress({
-          fromChain: sourceChain,
-          toChain: destChain,
-          destinationAddress: destinationAddress,
-          asset: coinMinimalDenom,
-          options: autoUnwrapIntoNative
-            ? {
-                shouldUnwrapIntoNative: autoUnwrapIntoNative,
-              }
-            : undefined,
-        })
-        .then((generatedAddress) => {
-          if (latestGenCacheKey.current === cacheKey)
-            setDepositAddress(generatedAddress);
-          depositAddressCache.current.set(cacheKey, generatedAddress);
-        })
-        .catch((e: any) => {
-          console.error("useDepositAddress > getDepositAddress:", e.message);
-        })
-        .finally(() => setIsLoading(false));
+    if (destinationAddress && shouldGenerate) {
+      if (!cacheHas(cacheKey)) {
+        const generateAddress = async () => {
+          try {
+            return await axelarAssetTransfer.getDepositAddress({
+              fromChain: sourceChain,
+              toChain: destChain,
+              destinationAddress: destinationAddress!,
+              asset: coinMinimalDenom,
+              options: { shouldUnwrapIntoNative },
+            });
+          } catch (e: unknown) {
+            if (e instanceof Error) {
+              console.error("useDepositAddress > generateAddress:", e.message);
+            } else {
+              console.error(
+                "useDepositAddress > generateAddress: Unknown error"
+              );
+            }
+            throw e;
+          }
+        };
+        cacheSet(cacheKey, fromPromise(generateAddress()));
+      }
+      return cacheGet(cacheKey);
     }
-    return null;
   }, [
-    environment,
-    destinationAddress,
+    axelarAssetTransfer,
     sourceChain,
     destChain,
-    coinMinimalDenom,
-    autoUnwrapIntoNative,
-    setIsLoading,
-  ]);
-
-  const doGen = useCallback(
```

Comment:
I removed `doGen` as it wasn't required. `doGen` wraps `generateAddress` which already returns a promise. Moreover, `generateAddress` already handles throw errors and doesn't return anything for successful calls. So line 82 here is unreachable since `address` is always going to be falsy.